google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java

```diff
@@ -0,0 +1,599 @@
+ *   <li>{@link ReadChangeStreamResponse.DataChange}s, that must be merged to a
+ *       ChangeStreamMutation.
+ *   <li>ChangeStreamRecord consumption events that reset the state machine for the next change
+ *       stream record.
+ * </ul>
+ *
+ * <p>The outputs are:
+ *
+ * <ul>
+ *   <li>Heartbeat records.
+ *   <li>CloseStream records.
+ *   <li>ChangeStreamMutation records.
+ * </ul>
+ *
+ * <p>Expected Usage:
+ *
+ * <pre>{@code
+ * ChangeStreamStateMachine changeStreamStateMachine = new ChangeStreamStateMachine<>(myChangeStreamRecordAdapter);
+ * while(responseIterator.hasNext()) {
+ *   ReadChangeStreamResponse response = responseIterator.next();
+ *   if (response.hasHeartbeat()) {
+ *     changeStreamStateMachine.handleHeartbeat(response.getHeartbeat());
+ *   } else if (response.hasCloseStream()) {
+ *     changeStreamStateMachine.handleCloseStream(response.getCloseStream());
+ *   } else {
+ *       changeStreamStateMachine.handleDataChange(response.getDataChange());
+ *   }
+ *   if (changeStreamStateMachine.hasCompleteChangeStreamRecord()) {
+ *       MyChangeStreamRecord = changeStreamStateMachine.consumeChangeStreamRecord();
+ *       // do something with the change stream record.
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <ChangeStreamRecordT> The type of row the adapter will build
+ */
+final class ChangeStreamStateMachine<ChangeStreamRecordT> {
+  private final ChangeStreamRecordBuilder<ChangeStreamRecordT> builder;
+  private State currentState;
+  // debug stats
+  private int numHeartbeats = 0;
+  private int numCloseStreams = 0;
+  private int numDataChanges = 0;
+  private int numNonCellMods = 0;
+  private int numCellChunks = 0; // 1 for non-chunked cell.
+  /**
+   * Expected total size of a chunked SetCell value, given by the {@link
+   * ReadChangeStreamResponse.MutationChunk.ChunkInfo}. This value should be the same for all chunks
+   * of a SetCell.
+   */
+  private int expectedTotalSizeOfChunkedSetCell = 0;
+
+  private int actualTotalSizeOfChunkedSetCell = 0;
+  private ChangeStreamRecordT completeChangeStreamRecord;
+
+  /**
+   * Initialize a new state machine that's ready for a new change stream record.
+   *
+   * @param builder The builder that will build the final change stream record.
+   */
+  ChangeStreamStateMachine(ChangeStreamRecordBuilder<ChangeStreamRecordT> builder) {
+    this.builder = builder;
+    reset();
+  }
+
+  /**
+   * Handle heartbeat events from the server.
+   *
+   * <dl>
+   *   <dt>Valid states:
+   *   <dd>{@link ChangeStreamStateMachine#AWAITING_NEW_STREAM_RECORD}
+   *   <dt>Resulting states:
+   *   <dd>{@link ChangeStreamStateMachine#AWAITING_STREAM_RECORD_CONSUME}
+   * </dl>
+   */
+  void handleHeartbeat(ReadChangeStreamResponse.Heartbeat heartbeat) {
+    try {
+      numHeartbeats++;
+      currentState = currentState.handleHeartbeat(heartbeat);
+    } catch (RuntimeException e) {
+      currentState = null;
```

Comment:
whats the difference between reset() and currentState = null? When an error occurs, wont the state machine be in a broken state? shouldnt you put currentState = ERROR and refuse any further input?

Reply:
Discussed offline. Added a new state ERROR which throws exception for any method called on it.